### PR TITLE
Support UNION_BY_NAME option in parquet_scan  read_parquet 

### PR DESCRIPTION
**Pull request by douenergy**
https://github.com/duckdb/duckdb/pull/4837 , https://github.com/duckdb/duckdb/issues/4720 , https://github.com/duckdb/duckdb/issues/4699 and https://github.com/duckdb/duckdb/pull/3238

We can make these queries much easier !
```sql
SELECT * FROM 'p1.parquet' UNION BY NAME
SELECT * FROM 'p2.parquet' UNION BY NAME
SELECT * FROM 'p3.parquet' 
```
to

```sql
SELECT *  FROM parquet_scan('p*.parquet' ,  union_by_name=TRUE);
```

* All Parquet files matched the glob pattern will union by their col name, If any Parquet file without some specified cols NULL will be generated for missing parquet cols. 

* As described in  https://github.com/duckdb/duckdb/issues/4720,  parquet_scan with union_by_name **isn't  reversible.**

If we have p1.parquet 

```
col1
a
b
c
```
and p2.parquet 

```
col1
1
2
3
```

```sql
INSERT INTO tbl SELECT * FROM read_parquet('*.parquet', FILENAME=TRUE, UNION_BY_NAME=TRUE);
```
then

```sql
COPY (SELECT * FROM tbl WHERE filename='p2.parquet') TO 'p2new.parquet';
```

`p2new.parquet`  and `p2.parquet ` will have different type. Maybe we should add some warning?

### CODE REVIEW

File: extension/parquet/parquet-extension.cpp [hunk 1]
```diff
@@ -303,11 +321,88 @@ class ParquetScanFunction {
...
+
+		for (idx_t file_idx = 0; file_idx < result->files.size(); ++file_idx) {
+			auto reader = make_shared<ParquetReader>(context, result->files[file_idx], parquet_options);
+			auto &col_names = reader->names;
+			auto &sql_types = reader->return_types;
+			D_ASSERT(col_names.size() == sql_types.size());
+
+			// union_col_names will exclude generated columns
+			// like filename, hivepartition etc.
+			for (idx_t col = 0; col <= reader->last_parquet_col; ++col) {
```
Comment (Mytherin) [inline extension/parquet/parquet-extension.cpp:356]: This code looks very similar to the code in `read_csv.cpp` - could we extract this into a separate class (e.g. `UnionByName` - similar to how we handle hive partitioning in the `HivePartitioning` class)?

File: extension/parquet/parquet_reader.cpp [hunk 2]
```diff
@@ -355,13 +356,29 @@ unique_ptr<ColumnReader> ParquetReader::CreateReader(const duckdb_parquet::forma
...
 	auto &root_struct_reader = (StructColumnReader &)*ret;
 
-	// add casts if required
-	for (auto &entry : cast_map) {
-		auto column_idx = entry.first;
-		auto &expected_type = entry.second;
-		auto child_reader = move(root_struct_reader.child_readers[column_idx]);
-		auto cast_reader = make_unique<CastColumnReader>(move(child_reader), expected_type);
-		root_struct_reader.child_readers[column_idx] = move(cast_reader);
+	if (!parquet_options.union_by_name) {
```
Comment (Mytherin) [inline extension/parquet/parquet_reader.cpp:359]: I'm not sure if modifying the Parquet Reader itself is necessary here - can't we handle the `union_by_name` in the `ParquetScanFunction` outside of the `ParquetReader` class? That should also allow for more code re-use between the CSV and Parquet readers for this functionality - by e.g. moving `SetNullUnionCols` to a shared `UnionByName` class.

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. Some comments below:

**Comment by Mytherin** [comment]
Could you also have a look at the failing CI?

File: extension/parquet/parquet-extension.cpp (hunk 1 above)
Comment (douenergy) [reply extension/parquet/parquet-extension.cpp:356]: Yes! That makes code much reusable

**Comment by douenergy** [comment]
> Could you also have a look at the failing CI?
parquet_union_by_name.test 's glob pattern also match the  files created by parquet_3989.test  😅

**Comment by danthegoodman1** [comment]
Very excited for this!

**Comment by danthegoodman1** [comment]
If a given file does not contain a column of interest (say we do `SELECT col_b FROM read_parquet... WHERE col_b IS NOT NULL`), will DuckDB be able to recognize from the metadata that the column does not exist and thus not read the file that does not contain the column?

**Comment by douenergy** [comment]
> If a given file does not contain a column of interest (say we do `SELECT col_b FROM read_parquet... WHERE col_b IS NOT NULL`), will DuckDB be able to recognize from the metadata that the column does not exist and thus not read the file that does not contain the column?

if you're saying , there are two parquet files 
p1.parquet 
```
colA   colB
  1.        a
  2         b
  3         c
```

p2.parquet 
```
colA   colC
  1.        x
  2         y
  3         z
```
For the query  `SELECT colB FROM read_parquet... WHERE colB IS NOT NULL`  just read only all metadata and completely skip p2.parquet 

Current Implement **cant** do that.  But I think we can improve it in another PR.

**Comment by danthegoodman1** [comment]
@douenergy sounds good :D I think it could be a worthwhile enhancement in the future!

**Comment by danthegoodman1** [comment]
@douenergy is there a way to make this behave like `UNION ALL BY NAME`?

**Comment by douenergy** [comment]
> @douenergy is there a way to make this behave like `UNION ALL BY NAME`?

Currently, parquet_scan with union_by_name just work like `UNION ALL BY NAME`.  
It allows duplicate row.

**Comment by danthegoodman1** [comment]
> Currently, parquet_scan with union_by_name just work like `UNION ALL BY NAME`. It allows duplicate row.

Perfect!

Does this issue still occur (it reading all files for something like a count even with a filter)? https://github.com/duckdb/duckdb/discussions/5790

**Comment by douenergy** [comment]
@danthegoodman1 
 I'think parquet_scan (with union_by_name) would help for your use case.
 Because there will no  UNION operator in your query plan(even you select from VIEW).

**Comment by danthegoodman1** [comment]
That'd be **AMAZING**. I can't get duckdb to build on my machine or in a github codespace to save my life so I will look forward to testing once a release comes out!

**Comment by douenergy** [comment]
Thanks @Mytherin  
I think I've now ready for another review.

1. Created an UnionByName template in common/union_by_name.hpp
2. Rename basecsvreader  "sql_types" , "col_names"  to "return_types" "names".  This make  basecsvreader  and parquetreader have same member names for same purpose.
3. fix the test.

**Comment by douenergy** [comment]
Unrelated CodeCov github action CI failed 
See https://github.com/codecov/codecov-action/issues/598

**Review by Mytherin — APPROVED** [review]
Thanks for the fixes! LGTM

**Pull request merged**